graphql/src/store/prefetch.rs

```diff
@@ -0,0 +1,870 @@
+            _ => None,
+        }
+    }
+
+    /// Return `true` if any of the entities matches this type condition
+    fn matches(&self, entities: &Vec<Node>) -> bool {
+        use TypeCondition::*;
+        match self {
+            Any => true,
+            On(name) => entities.iter().any(|entity| entity.typename() == name),
+        }
+    }
+
+    /// Return the type that matches this condition; for `Any`, use `object_type`
+    fn matching_type<'a>(
+        &self,
+        schema: &'a s::Document,
+        object_type: &'a ObjectOrInterface<'a>,
+    ) -> Option<ObjectOrInterface<'a>> {
+        use TypeCondition::*;
+
+        match self {
+            Any => Some(object_type.to_owned()),
+            On(name) => object_or_interface_by_name(schema, name),
+        }
+    }
+}
+
+impl From<Option<q::TypeCondition>> for TypeCondition {
+    fn from(cond: Option<q::TypeCondition>) -> Self {
+        match cond {
+            None => TypeCondition::Any,
+            Some(q::TypeCondition::On(name)) => TypeCondition::On(name),
+        }
+    }
+}
+
+/// Intermediate data structure to hold the results of prefetching entities
+/// and their nested associations. For each association of `entity`, `children`
+/// has an entry mapping the response key to the list of nodes.
+#[derive(Debug, Clone)]
+struct Node {
+    entity: Entity,
+    /// We are using an `Rc` here for two reasons: it allows us to defer
+    /// copying objects until the end, when converting to `q::Value` forces
+    /// us to copy any child that is referenced by multiple parents. It also
+    /// makes it possible to avoid unnecessary copying of a child that is
+    /// referenced by only one parent - without the `Rc` we would have to copy
+    /// since we do not know that only one parent uses it.
+    children: BTreeMap<String, Vec<Rc<Node>>>,
+}
+
+impl From<Entity> for Node {
+    fn from(entity: Entity) -> Self {
+        Node {
+            entity,
+            children: BTreeMap::default(),
+        }
+    }
+}
+
+/// Convert a list of nodes into a `q::Value::List` where each node has also
+/// been converted to a `q::Value`
+fn node_list_as_value(nodes: Vec<Rc<Node>>) -> q::Value {
+    q::Value::List(
+        nodes
+            .into_iter()
+            .map(|node| Rc::try_unwrap(node).unwrap_or_else(|rc| rc.as_ref().clone()))
+            .map(|node| node.into())
+            .collect(),
+    )
+}
+
+/// We pass the root node of the result around as a vec of nodes, not as
+/// a single node so that we can use the same functions on interior node
+/// lists which are the result of querying the database. The root list
+/// consists of exactly one entry, and that entry has an empty
+/// (not even a `__typename`) entity.
+///
+/// That distinguishes it from both the result of a query that matches
+/// nothing (an empty `Vec`), and a result that finds just one entity
+/// (the entity is not completely empty)
```

Comment:
A cleaner way would be for `Node` to be an enum and to have a `Node::Root` variant for the root.

Reply:
I thought about doing that, but it would complicate the code quite a bit, since the root node and interior nodes are treated exactly the same, except for one place where we decide whether we need to window the query or not (in `fetch`) And since the data structure is completely internal to `prefetch`, I don't think it will lead to a lot of confusion.

Reply:
Yeah, I thought so. Just wanted to bring it up.